### PR TITLE
Migrate Child Block Test to Playwright

### PR DESCRIPTION
**Pull request by pooja-muchandikar**
## What?

Part of https://github.com/WordPress/gutenberg/issues/38851.
Migrate[ child-blocks.test.js ](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/specs/editor/plugins/child-blocks.test.js) to its Playwright version.
## Why?

Part of https://github.com/WordPress/gutenberg/issues/38851.
## How?

See[ MIGRATION.md](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/MIGRATION.md) for migration steps.
## Testing Instructions

Run` npm run test:e2e:playwright test/e2e/specs/editor/plugins/child-blocks.spec.js`

### CODE REVIEW

**Comment by pooja-muchandikar** [comment]
Hi @kevin940726 @Mamaduka,

I hope you are doing great!

Could you help review this PR?

Thanks!

**Review by Mamaduka — COMMENTED** [review]
Thank you, @pooja-muchandikar!

I think some of the locators can be improved. Especially ones that use classes instead of role-based selectors.

Recently, I did a similar migration (#51677); you can check the [`inner-blocks-allowed-blocks.spec.js`](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/specs/editor/plugins/inner-blocks-allowed-blocks.spec.js) file and use it as a blueprint.

**Comment by pooja-muchandikar** [comment]
Hi @Mamaduka,

Thanks for sharing the feedbacks! All the feedbacks are addressed please review. 

Thanks!

**Comment by Mamaduka** [comment]
Thanks, @pooja-muchandikar!

While changes look good, the two tests now check a different thing. They should test a block appender based on the original tests. Let's ensure that migrate tests are still checking for the same thing.

## Affected tests

* `shows up in a parent block`
* `display in a parent block with allowedItems`.


## Block appender
CleanShot 2023-10-13 at 13 25 30

**Comment by Mamaduka** [comment]
@pooja-muchandikar, let me know if you have any additional questions.

**Comment by pooja-muchandikar** [comment]
Hi @Mamaduka, 

I have addressed the feedback and pushed the changes. Please review and let me know if there are any further suggestions. 

Thanks

File: test/e2e/specs/editor/plugins/child-blocks.spec.js [hunk 1]
```diff
@@ -36,11 +36,9 @@ test.describe( 'Child Blocks', () => {
...
 		} );
 
-		await page
-			.getByRole( 'document', {
-				name: 'Block: Child Blocks Unrestricted Parent',
-			} )
-			.click();
+		await page.click(
+			'[data-type="test/child-blocks-unrestricted-parent"] .block-editor-default-block-appender'
+		);
```
Comment (Mamaduka) [inline test/e2e/specs/editor/plugins/child-blocks.spec.js]: We should try and use [accessible selectors](https://github.com/WordPress/gutenberg/blob/a3f6e952db5c60e9197d0881b832d0efde186573/docs/contributors/code/e2e/README.md#use-accessible-selectors) when possible.

Example:

```js
await page
				.getByRole( 'document', {
					name: 'Block: Child Blocks Unrestricted Parent',
				} )
				.getByRole( 'button', {
					name: 'Add block',
				} )
				.click();
```

File: test/e2e/specs/editor/plugins/child-blocks.spec.js [hunk 2]
```diff
@@ -63,11 +61,9 @@ test.describe( 'Child Blocks', () => {
...
 		} );
 
-		await page
-			.getByRole( 'document', {
-				name: 'Block: Child Blocks Restricted Parent',
-			} )
-			.click();
+		await page.click(
+			'[data-type="test/child-blocks-restricted-parent"] .block-editor-default-block-appender'
+		);
```
Comment (Mamaduka) [inline test/e2e/specs/editor/plugins/child-blocks.spec.js]: Same here.

Comment (pooja-muchandikar) [reply test/e2e/specs/editor/plugins/child-blocks.spec.js]: Hi @Mamaduka, I tried to use the solution here but it is not working as expected. 

`await page
				.getByRole( 'document', {
					name: 'Block: Child Blocks Restricted Parent',
				} )
				.getByRole( 'button', {
					name: 'Add block',
				} )
				.click();`

File: test/e2e/specs/editor/plugins/child-blocks.spec.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		await editor.insertBlock( {
+			name: 'test/child-blocks-unrestricted-parent',
+		} );
+
+		await page
+			.getByRole( 'document', {
+				name: 'Block: Child Blocks Unrestricted Parent',
+			} )
+			.getByRole( 'button', {
+				name: 'Add block',
```
Comment (Mamaduka) [inline test/e2e/specs/editor/plugins/child-blocks.spec.js]:
```suggestion
				name: 'Add default block',
```

I just noticed that the original test was clicking on the default appender.

File: test/e2e/specs/editor/plugins/child-blocks.spec.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+		const blockInserter = page
+			.getByRole( 'toolbar', { name: 'Document tools' } )
+			.getByRole( 'button', { name: 'Toggle block inserter' } );
+		const blockLibrary = page.getByRole( 'region', {
+			name: 'Block Library',
+		} );
+
+		await blockInserter.click();
+		await expect( blockLibrary ).toBeVisible();
```
Comment (Mamaduka) [inline test/e2e/specs/editor/plugins/child-blocks.spec.js:56]:
```suggestion
		await expect( blockLibrary ).toBeVisible();
		await expect( blockLibrary.getByRole( 'option' ) ).toContainText( [
			'Child Blocks Child',
		] );
```

This assertion was missing from the migration.

File: test/e2e/specs/editor/plugins/child-blocks.spec.js (hunk 2 above)
Comment (Mamaduka) [reply test/e2e/specs/editor/plugins/child-blocks.spec.js]: I think this should work now:

```js
await page
			.getByRole( 'document', {
				name: 'Block: Child Blocks Restricted Parent',
			} )
			.getByRole( 'button', {
				name: 'Add default block',
			} )
			.click();
```

**Review by Mamaduka — COMMENTED** [review]
I just ran the original test in headed mode and adjusted my suggestions. The tests should be passing now.

**Comment by pooja-muchandikar** [comment]
Hi @Mamaduka 

Feedbacks addressed.

File: test/e2e/specs/editor/plugins/child-blocks.spec.js [hunk 5]
```diff
@@ -29,6 +29,9 @@ test.describe( 'Child Blocks', () => {
 		expect( blockLibrary.getByRole( 'option' ) ).not.toContain( [
 			'Child Blocks Child',
 		] );
+		expect(
+			await blockLibrary.getByRole( 'option' ).count()
+		).toBeGreaterThan( 10 );
```
Comment (Mamaduka) [inline test/e2e/specs/editor/plugins/child-blocks.spec.js]: I think this addition wasn't necessary here.

**Review by Mamaduka — APPROVED** [review]
Thank you, @pooja-muchandikar! One last nitpick and I think we're good to merge once CI checks are green.

File: test/e2e/specs/editor/plugins/child-blocks.spec.js (hunk 5 above)
Comment (pooja-muchandikar) [reply test/e2e/specs/editor/plugins/child-blocks.spec.js]: Done

**Pull request merged**